File: bookstore/spring-backend/src/main/kotlin/masha/bookstore/services/ReviewService.kt
package masha.bookstore.services

import masha.bookstore.core.BaseService
import masha.bookstore.models.*
import masha.bookstore.interfaces.IReviewService
import masha.bookstore.repositories.ReviewRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ReviewService(
    @Autowired override val repository: ReviewRepository
) : BaseService<Review, ReviewRepository>(repository), IReviewService {

    override fun create(book: Book, author: String, content: String): Review {
        val review = Review(book, author, content)
        repository.save(review)
        return review
    }

    override fun change(id: Int, content: String): Review {
        val review: Review = getOne(id)
        review.content = content
        repository.save(review)
        return review
    }
}

File: bookstore/mern/services/review.service.js
import ReviewModel from '../models/review.model.js'

export default class ReviewService {
  async getOne(_id) {
    return ReviewModel.findById(_id).lean()
  }

  async create(author, content) {
    const review = await new ReviewModel({ author, content })
    await review.save()
    return review
  }

  async change(_id, content) {
    const review = await ReviewModel.findByIdAndUpdate(_id, {
      $set: { content },
    }).lean()
    return { ...review, content }
  }

  async delete(_id) {
    await ReviewModel.findByIdAndRemove(_id)
    return _id
  }
}
